src/components/lib/component/arch/x86_64/cos_component.h

```diff
@@ -139,22 +140,36 @@ cos_print(char *s, int len)
 	call_cap(PRINT_CAP_TEMP, (word_t) s, len, 0, 0);
 }
 
+typedef int (*callgate_fn_t)(word_t p0, word_t p1, word_t p2, word_t p3, word_t *r1, word_t *r2);
+
 static inline int
-cos_sinv(u32_t sinv, word_t arg1, word_t arg2, word_t arg3, word_t arg4)
+cos_sinv(word_t arg1, word_t arg2, word_t arg3, word_t arg4)
 {
-	return call_cap_op(sinv, 0, arg1, arg2, arg3, arg4);
+	register struct usr_inv_cap *uc __asm__ ("rax");
+	register callgate_fn_t callgate __asm__ ("r11");  
```

Comment:
> I don't know where these values are being put in the registers, but now I see that there are other C functions executing in the mean time. That means that this code is assuming that the registers `a` and `11` aren't used in any of the intervening logic. I don't see how this assumption could hold, but I could be missing something.

https://github.com/gwsystems/composite/blob/f580b71313c1841449916ffe2ef3b7a22bb2f368/src/components/lib/component/c_stub.c#L8
Previously we use this macro to set the `rax` as the `uc`. @evanstella seems to put this logic all into `cos_sinv`. I think the previous one above can work is because that we use `jmp` directly call into the stub function(using `a` register as `uc`). Then, it calls `cos_sinv `, which uses `call` rather than `jmp`. There should be some calling conventions & compiler's decisions that all decide whether `ax` and `r11` will be used as what we want. I searched that it seems the compiler doesn't guarantee this.